src/Parallel/Spinlock.hpp

```diff
@@ -0,0 +1,58 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <atomic>
+
+namespace Parallel {
+/*!
+ * \brief A simple spinlock implemented in `std::atomic`s
+ *
+ * Implementation adapted from https://rigtorp.se/spinlock/
+ */
+class Spinlock {
+ public:
+  Spinlock() = default;
+  Spinlock(const Spinlock&) = delete;
+  Spinlock& operator=(const Spinlock&) = delete;
+  Spinlock(Spinlock&&) = delete;
+  Spinlock& operator=(Spinlock&&) = delete;
+  ~Spinlock() = default;
+
+  /// \brief Acquire the lock. Will block until the lock is acquire.
+  void lock() {
+    for (;;) {
+      // Optimistically assume the lock is free on the first try
+      if (not lock_.exchange(true, std::memory_order_acquire)) {
+        return;
+      }
+      // Wait for lock to be released without generating cache misses
+      while (lock_.load(std::memory_order_relaxed)) {
+        // Issue X86 PAUSE or ARM YIELD instruction to reduce contention between
```

Comment:
is there a reason to keep this comment?

Reply:
I've added a comment explaining that this is something we might want to experiment with in the future, so I'm keeping it here as "knowledge"